HermesProxy/World/Client/PacketHandlers/ItemHandler.cs

```diff
@@ -214,5 +214,17 @@ void HandleItemEnchantTimeUpdate(WorldPacket packet)
             enchant.OwnerGuid = packet.ReadGuid().To128(GetSession().GameState);
             SendPacketToClient(enchant);
         }
+
+        [PacketHandler(Opcode.SMSG_ENCHANTMENT_LOG)]
+        void HandleEnchantmentLog(WorldPacket packet)
+        {
+            EnchantmentLog enchantment = new EnchantmentLog();
+            enchantment.Caster = packet.ReadPackedGuid().To128(GetSession().GameState);
+            enchantment.Owner = enchantment.Caster;
+            enchantment.ItemGUID = packet.ReadPackedGuid().To128(GetSession().GameState);
+            enchantment.ItemID = (int)packet.ReadUInt32();
+            enchantment.Enchantment = (int)packet.ReadUInt32();
+            SendPacketToClient(enchantment);
```

Comment:
The structure you are reading is invalid for 1.12. Hermes will crash when it receives the packet. Always check both versions of mangos when implementing packets.

This is the vanilla structure.
```
data << ObjectGuid(targetGuid);
data << ObjectGuid(casterGuid);
data << uint32(itemId);
data << uint32(spellId);
data << uint8(0);
```

This is the tbc structure.
```
data << GetPackGUID();
data << targetGuid.WriteAsPacked();
data << uint32(itemEntry);
data << uint32(enchantId);
```

Reply:
I don't have 1.12 so I can't check that unfortunately :( Thank you!

Reply:
This is not what the guids mean, at least in vanilla! The first guid is the player who owns the item, and the second guid is the player who applied the enchant. If the second guid is empty, it means the enchant has faded. Also the byte at the end is a bool and indicates whether to show affiliation in the combat log or not. When sent to yourself it will be false. When sent to nearby players it will be true.

Reply:
Is it even possible to implement that properly, without (breaking) serverside changes then?
The Client expects one of these GUIDs to be the item GUID

Reply:
```cs
        void HandleEnchantmentLog(WorldPacket packet)
        {
            EnchantmentLog enchantment = new EnchantmentLog();
            if (LegacyVersion.AddedInVersion(ClientVersionBuild.V2_0_1_6180))
            {
                enchantment.Caster = packet.ReadPackedGuid().To128(GetSession().GameState);
                enchantment.Owner = packet.ReadPackedGuid().To128(GetSession().GameState);
            }
            else
            {
                enchantment.Owner = packet.ReadPackedGuid().To128(GetSession().GameState);
                enchantment.Caster = packet.ReadPackedGuid().To128(GetSession().GameState);
            }
            enchantment.ItemGUID = new WowGuid128();
            enchantment.ItemID = (int)packet.ReadUInt32();
            enchantment.Enchantment = (int)packet.ReadUInt32();
            SendPacketToClient(enchantment);
        }
```

Reply:
How do I go through known items?

Reply:
```
        void HandleEnchantmentLog(WorldPacket packet)
        {
            EnchantmentLog enchantment = new EnchantmentLog();
            if (LegacyVersion.AddedInVersion(ClientVersionBuild.V2_0_1_6180))
            {
                enchantment.Caster = packet.ReadPackedGuid().To128(GetSession().GameState);
                enchantment.Owner = packet.ReadPackedGuid().To128(GetSession().GameState);
            }
            else
            {
                enchantment.Owner = packet.ReadPackedGuid().To128(GetSession().GameState);
                enchantment.Caster = packet.ReadPackedGuid().To128(GetSession().GameState);
            }
            enchantment.ItemID = (int)packet.ReadUInt32();
            foreach (var obj in GetSession().GameState.ObjectCacheModern)
            {
                if (obj.Key.GetObjectType() == ObjectType.Item && obj.Value.GetUpdateField<uint>(ObjectField.OBJECT_FIELD_ENTRY).Equals(enchantment.ItemID))
                {
                    enchantment.ItemGUID = obj.Key;
                    break;
                }
            }
            enchantment.Enchantment = (int)packet.ReadUInt32();
            SendPacketToClient(enchantment);
        }
```

Reply:
There already is an existing method that returns item in inventory or equipped.

Reply:
```cs
        void HandleEnchantmentLog(WorldPacket packet)
        {
            EnchantmentLog enchantment = new EnchantmentLog();
            if (LegacyVersion.AddedInVersion(ClientVersionBuild.V2_0_1_6180))
            {
                enchantment.Caster = packet.ReadPackedGuid().To128(GetSession().GameState);
                enchantment.Owner = packet.ReadPackedGuid().To128(GetSession().GameState);
            }
            else
            {
                enchantment.Owner = packet.ReadPackedGuid().To128(GetSession().GameState);
                enchantment.Caster = packet.ReadPackedGuid().To128(GetSession().GameState);
            }
            enchantment.ItemID = (int)packet.ReadUInt32();
            enchantment.ItemGUID = new WowGuid128();
            ObjectUpdate updateData = new ObjectUpdate(GetSession().GameState.CurrentPlayerGuid, UpdateTypeModern.Values, GetSession());
            ActivePlayerData activeData = updateData.ActivePlayerData;
            for (int i = 0; i < 23; i++)
                {
                    if (activeData.InvSlots[i] != null)
                        if (activeData.InvSlots[i].GetEntry().Equals(enchantment.ItemID))
                        {
                            enchantment.ItemGUID = activeData.InvSlots[i];
                            break;
                        }
                }
            enchantment.Enchantment = (int)packet.ReadUInt32();
            SendPacketToClient(enchantment);
        }
```